### PR TITLE
Restructure Laser Profiles

### PR DESCRIPTION
**Pull request by MKirchen**
This PR restructures the various laser profiles that are available in the code.

In particular, this PR introduces separate transverse and longitudinal laser profiles which can be arbitrarily combined to construct a laser that is valid under the paraxial approximation. Most of the existing laser profiles have been rewritten and are now expressed as a combination of such profiles.

Please merge this PR before merging #561.

### CODE REVIEW

File: fbpic/lpa_utils/laser/transverse_laser_profiles.py [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+        """
+        # Initialize propagation direction and mark the profile as GPU capable
+        LaserTransverseProfile.__init__(self, propagation_direction,
+                                        gpu_capable=True)
+
+        # Wavevector and Rayleigh length
+        k0 = 2 * np.pi / lambda0
+        zr = 0.5 * k0 * waist ** 2
+        # Store the parameters
+        self.k0 = k0
```
Comment (RemiLehe) [inline fbpic/lpa_utils/laser/transverse_laser_profiles.py:115]: It seems that we do not need to store `k0` here.
```suggestion
        self.k0 = k0
```

File: fbpic/lpa_utils/laser/transverse_laser_profiles.py [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+            scaled_amplitude *= 2**.5
+        # Store the parameters
+        if m < 0 or type(m) is not int:
+            raise ValueError("m should be an integer positive number.")
+        self.p = p
+        self.m = m
+        self.scaled_amplitude = scaled_amplitude
+        self.laguerre_pm = genlaguerre(self.p, self.m) # Laguerre polynomial
+        self.theta0 = theta0
+        self.k0 = k0
```
Comment (RemiLehe) [inline fbpic/lpa_utils/laser/transverse_laser_profiles.py:246]: It seems that we do not need to store `k0`.
```suggestion
        self.k0 = k0
```

File: fbpic/lpa_utils/laser/transverse_laser_profiles.py [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+        k0 = 2*np.pi/lambda0
+        zr = 0.5*k0*waist**2
+        # Scaling factor, so that the pulse energy is independent of p and m.
+        scaled_amplitude = np.sqrt( factorial(p)/factorial(abs(m)+p) )
+        # Store the parameters
+        self.p = p
+        self.m = m
+        self.scaled_amplitude = scaled_amplitude
+        self.laguerre_pm = genlaguerre(self.p, abs(m)) # Laguerre polynomial
+        self.k0 = k0
```
Comment (RemiLehe) [inline fbpic/lpa_utils/laser/transverse_laser_profiles.py:361]:
```suggestion
```

File: fbpic/lpa_utils/laser/transverse_laser_profiles.py [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+
+        # Ensure that N is an integer
+        self.N = int(round(N))
+        # Calculate effective waist of the Laguerre-Gauss modes, at focus
+        self.w_foc = w0*(self.N+1)**.5
+        # Wavevector and Rayleigh length
+        k0 = 2 * np.pi / lambda0
+        zr = 0.5 * k0 * self.w_foc**2
+        # Store laser parameters
+        self.k0 = k0
```
Comment (RemiLehe) [inline fbpic/lpa_utils/laser/transverse_laser_profiles.py:473]:
```suggestion
```

File: fbpic/lpa_utils/laser/transverse_laser_profiles.py [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+                L1 = L
+                L = (((2*n -1) - scaled_radius_squared) * L1 - (n - 1) * L2) / n
+            # Add to the sum, including the term for the additional Gouy phase
+            laguerre_sum += self.cn[n] * np.exp( - (2j* n) * psi ) * L
+
+        # Final profile: multiply by n-independent propagation factors
+        exp_argument = - (x**2 + y**2) / (self.w_foc**2 * diffract_factor)
+        profile = laguerre_sum * np.exp( exp_argument ) / diffract_factor
+
+        return profile
```
Comment (RemiLehe) [inline fbpic/lpa_utils/laser/transverse_laser_profiles.py:517]: Could you fix the line ending here? It seems that `git` does not like it.

**Review by RemiLehe — COMMENTED** [review]
Great PR! Thanks a lot!
I added a few minor comments.

**Comment by MKirchen** [comment]
I implemented the requested changes. However, I keep storing the wavenumber in the longitudinal and transverse profiles as the Paraxial Approximation Laser (#561) needs it to check for consistency.

**Review by RemiLehe — APPROVED** [review]

**Pull request merged**